python/cornserve/services/gateway/app/manager.py

```diff
@@ -109,66 +109,93 @@ def __init__(self, task_manager: TaskManager) -> None:
             while True:
                 app_id = f"app-{uuid.uuid4().hex}"
                 if app_id not in self.app_states:
                     break
 
-            self.app_states[app_id] = AppState.NOT_READY
-        span.set_attribute("app_manager.register_app.app_id", app_id)
+        span.set_attribute("app_manager.validate_and_create_app.app_id", app_id)
 
         try:
-            # Load and validate the app
             module = load_module_from_source(source_code, app_id)
             app_classes = validate_app_module(module)
-            tasks = discover_unit_tasks(app_classes.config_cls.tasks.values())
+            tasks = expand_tasks_into_unit_tasks(app_classes.config_cls.tasks.values())
+            task_names = [t.execution_descriptor.create_executor_name().lower() for t in tasks]
+        except (ImportError, ValueError) as e:
+            raise ValueError(f"App source code validation failed: {e}") from e
+
+        async with self.app_lock:
+            self.apps[app_id] = AppDefinition(
+                app_id=app_id,
+                module=module,
+                classes=app_classes,
+                source_code=source_code,
+                tasks=tasks,
+            )
+            self.app_states[app_id] = AppState.NOT_READY
+
+        return app_id, task_names
+
+    @tracer.start_as_current_span(name="AppManager.deploy_app_tasks")
+    async def deploy_app_tasks(self, app_id: str) -> dict[str, str]:
+        """Deploy tasks for an app and return final result.
 
-            # Deploy all unit tasks discovered
-            await self.task_manager.declare_used(tasks)
+        Args:
+            app_id: The App's ID
+
+        Returns:
+            dict: Final status with either success or failure information
+
+        Raises:
+            RuntimeError: If task deployment gets any Exception
+        """
+        span = trace.get_current_span()
+        span.set_attribute("app_manager.deploy_app_tasks.app_id", app_id)
 
-            # Update app state and store app definition
+        tasks_to_deploy = []
+        try:
+            async with self.app_lock:
+                app_def = self.apps[app_id]
+
+            tasks_to_deploy = app_def.tasks
+
+            # Deploy unit tasks
+            await self.task_manager.declare_used(tasks_to_deploy)
+
+            # Update app state
             async with self.app_lock:
                 self.app_states[app_id] = AppState.READY
-                self.apps[app_id] = AppDefinition(
-                    app_id=app_id,
-                    module=module,
-                    classes=app_classes,
-                    source_code=source_code,
-                )
 
-            logger.info("Successfully registered app '%s'", app_id)
+            logger.info("Successfully deployed %s tasks for app '%s'.", len(tasks_to_deploy), app_id)
 
-            return app_id
+            return {"status": "ready", "message": f"Successfully deployed {len(tasks_to_deploy)} tasks"}
```

Comment:
Across the router and app manager, please strongly type returned & streamed data with `pydantic.BaseModel`, and also import them from the CLI. Dictionaries with literal string keys are error-prone and not type-safe.